Python/ceval.c

```diff
@@ -3688,6 +3689,99 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
+            _PyLoadMethodCache *cache = (_PyLoadMethodCache *)next_instr;
+            PyObject *owner = TOP();
+            PyTypeObject *cls = Py_TYPE(owner);
+            uint32_t type_version = read_u32(cache->type_version);
+            DEOPT_IF(cls->tp_version_tag != type_version, LOAD_ATTR);
+            assert(type_version != 0);
+            STAT_INC(LOAD_ATTR, hit);
+
+            PyObject *name = GETITEM(names, oparg >> 1);
+            PyObject *getattribute = ((PyHeapTypeObject *)cls)->_spec_cache.getattribute;
+            PyObject *getattr = ((PyHeapTypeObject *)cls)->_spec_cache.getattr;
+            PyObject *res = getattribute ?
+                _Py_call_attribute(owner, getattribute, name)
+                : PyObject_GenericGetAttr(owner, name);
+            if (res == NULL &&
+                getattr &&
+                PyErr_ExceptionMatches(PyExc_AttributeError)) {
+                PyErr_Clear();
+                res = _Py_call_attribute(owner, getattr, name);
+            }
+            if (res == NULL) {
+                goto error;
+            }
+            SET_TOP(NULL);
+            STACK_GROW((oparg & 1));
+            SET_TOP(res);
+            Py_DECREF(owner);
+            JUMPBY(INLINE_CACHE_ENTRIES_LOAD_ATTR);
+            DISPATCH();
+        }
+
+        TARGET(LOAD_ATTR_GETATTRIBUTE_GETATTR_PY_OVERRIDDEN) {
+            assert(cframe.use_tracing == 0);
+            DEOPT_IF(tstate->interp->eval_frame, LOAD_ATTR);
+            _PyLoadMethodCache *cache = (_PyLoadMethodCache *)next_instr;
+            PyObject *owner = TOP();
+            PyTypeObject *cls = Py_TYPE(owner);
+            uint32_t type_version = read_u32(cache->type_version);
+            DEOPT_IF(cls->tp_version_tag != type_version, LOAD_ATTR);
+            assert(type_version != 0);
+            PyObject *getattr = ((PyHeapTypeObject *)cls)->_spec_cache.getattr;
+            assert(Py_IS_TYPE(getattr, &PyFunction_Type));
+            PyFunctionObject *f = (PyFunctionObject *)getattr;
+            uint32_t func_version = read_u32(cache->keys_version);
+            assert(func_version != 0);
+            DEOPT_IF(f->func_version != func_version, LOAD_ATTR);
+            PyCodeObject *code = (PyCodeObject *)f->func_code;
+            assert(code->co_argcount == 2);
+            STAT_INC(LOAD_ATTR, hit);
+
+            PyObject *name = GETITEM(names, oparg >> 1);
+            PyObject *getattribute = ((PyHeapTypeObject *)cls)->_spec_cache.getattribute;
+            /* Possible todo: inlining this call frame? */
+            PyObject *res = getattribute ?
+                _Py_call_attribute(owner, getattribute, name)
+                : PyObject_GenericGetAttr(owner, name);
+            if (res) {
+                SET_TOP(NULL);
+                STACK_GROW((oparg & 1));
+                SET_TOP(res);
+                Py_DECREF(owner);
+                JUMPBY(INLINE_CACHE_ENTRIES_LOAD_ATTR);
+                DISPATCH();
+            }
```

Comment:
I'm aware that specialized opcodes should have as few branches as possible. But these two opcodes need to respect the `getattro` semantics.

I could move these out to a function, but the branch would still be there.